### ISSUE TITLE
Asynchronous font initialization

### ISSUE THREAD

**Issue by mirh** [Psychtoolbox-3/Psychtoolbox-3#786]
**Describe the bug**
While investigating [high](https://mathworks.com/matlabcentral/answers/437301-why-does-matlab-crash-when-using-the-screen-command-from-the-third-party-psychtoolbox) and [low](https://gitlab.freedesktop.org/fontconfig/fontconfig/-/issues/334) the ways of #579 I came to notice
https://gitlab.gnome.org/GNOME/gimp/-/commit/30b73125f87b19e3ae322b9fc5bacc9f0b8ef3c1
https://gitlab.gnome.org/GNOME/gimp/-/commit/2484dec7d55e75499712d913921d10902d326332
Now, I'm not saying that you should have some fancy game-like loading animation too, but to the very least whatever you or gstreamer do with fonts should not get to affect even the whole matlab GUI. 

**To Reproduce**
1. Have a fresh system
2. Do something with DrawText

**Expected behavior**
Fontconfig not taking an atrociously long time to build its cache to begin with, is not something for this project to fix - but regardless it should never happen that *everything* just [hangs](https://learn.microsoft.com/en-us/windows/win32/win7appqual/preventing-hangs-in-windows-applications). 

**Desktop:**
 - OS: Windows-10 (21H2)
 - Runtime: Matlab R2022b
 - Hardware anything?
 - Version Psychtoolbox 3.0.18

**Additional context**
This would also kinda address #436.

**Comment by kleinerm** [comment]
A lot of things should never happen, e.g., that a project of the complexity and size of PTB does receive essentially no financial support by the vast majority of its users, so is constantly underfunded. But such is life, and people sooner or later will get what they pay for.

This issue is a luxury problem which won't ever get improved unless somebody puts a large amount of work into it or pays a lot of money for the work time to improve this. We can't draw text before the fontconfig caches are built, so DrawText has to block until then. Adding complex visual feedback during cache rebuilt would require a lot of work, e.g., multi-threading inside the drawtext plugin, just to run some OpenGL rendered feedback animation.

The PTB installers/updaters trigger a hidden DrawText call to try to trigger a cache rebuild during post-installation, while a text message is displayed in the command window to notify users to wait patiently for seconds or minutes. That's as good as it gets.
Closing.

**Comment by mirh** [comment]
You mean the message in `WelcomeSplash.ppm`?
That's already not a bad *format* to warn the user that they might have to wait a while, but atm it actually just says "Please wait a few seconds while we test your graphics drivers". 
That's 100% fine and dandy when you are \*in fact\* only doing that, but I have seen awful laptops taking a good half a minute to complete their first cache run. You need something more specific in this case (hopefully also with a `help DrawTextPlugin` note to learn what to do whenever whatever glitch happens)

And regardless of anything, I think you are doing yourself a disservice by treating "can't be bothered" issues with the same ominous label of "rejected" ones.

**Comment by kleinerm** [comment]
The post install routine takes care of this issue in the common cases, that's good enough for me.
You are welcome to contribute all the necessary work yourself and send a high quality, well tested, easy to review and merge pull request, if this matters to you so much. But for me this is not a priority at all, as long as nobody pays for the needed work time. I don't have the funding to deal with fringe problems for free anymore, I don't even have the funding for many important problems, so I pick what I'd like to work on in my "spare time". And this is not it. The wontfix label is a clear statement here to set the proper expectations.

**Comment by mirh** [comment]
*`PsychtoolboxPostInstallRoutine.m`* doesn't take care of much tbh, because I assure you that I have seen an entire laboratory getting pissed that "the first run is always the longest one" (first, not in an absolute sense as in "ever", but meaning the first of the day or the first of the "work session"). 
Hopefully these kind of glitches improved now with newer gstreamer, though even if we were just talking about "legit expected behaviour" (like, I don't know, could cache rebuilding occasionally happen after some major OS update?) the normal unassuming splash screen getting shoehorned to also cover for this makes everything feel more like cheap quirky voodoo than solid reproducible software. 

If actually addressing the causes is too much effort for too little gain then (the GIMP guys made it seem like a cakewalk, lol), could at least another panel be shown before the rebuild triggers? Even just a simple black screen with two sentences would suffice (after all, it is not a screen that should be encountered often.. and if this is not the case, then we'd have a bigger problem). 
Assuming there is even a way to know this ahead of time.

p.s. *Closing* is already a clear statement of intents, and you are entitled not to want clutter in your issues/priorities list.
But *wontfix* is actually going one other step further if you ask me. Maybe it's still not as harsh as *invalid*, but it still doesn't really give much room for "rehabilitation" (whether this means future you in 10 years, or some other keen chap that might land here).
Low-priority, long-term, or simply TODO would seem much clearer (and easier to discern) ways to signal that you like the idea "at least in theory", and that it's something still genuinely up for grabs.

**Comment by dcnieho** [comment]
If it triggers every day, there might be an issue with the fontconfig
cache. I've run into that, deleting the cache did the trick

**Comment by kleinerm** [comment]
> PsychtoolboxPostInstallRoutine.m doesn't take care of much tbh, because I assure you that I have seen an entire laboratory getting pissed that "the first run is always the longest one" (first, not in an absolute sense as in "ever", but meaning the first of the day or the first of the "work session").

Hyberbole much? And that the first run takes a few seconds longer is a) by design, and b) has nothing to do with text rendering, as a cache rebuild would only happen at all on first invocation of a Screen 'DrawText' or 'TextBounds' function. Maybe if that entire lab is getting pissed so much about having to wait a few extra seconds once a day, in exchange for being able to use some free high quality software that is free to them, but absolutely not free to make and maintain, they should have filed a **paid** support request about the issue to help us and themselves, because surely a whole lab can afford to buy a PTB membership to give something back.

> Hopefully these kind of glitches improved now with newer gstreamer, though even if we were just talking about "legit expected behaviour" (like, I don't know, could cache rebuilding occasionally happen after some major OS update?) the normal unassuming splash screen getting shoehorned to also cover for this makes everything feel more like cheap quirky voodoo than solid reproducible software.

First of all, I don't care about such little "cheap quirky voodoo" as you state it in this nice entitled sounding way, unless somebody provides me with the funding and resources to be able to care about it. There is way bigger fish to fry, and I have more than enough - largely unpaid - work to do, to deal with the stuff that really matters for reproducible high quality science or the survival of the project. Solid reproducible software needs solid funding from its users. If our users don't want to support us, they are not entitled to anything and they get what they get.

> If actually addressing the causes is too much effort for too little gain then (the GIMP guys made it seem like a cakewalk, lol), could at least another panel be shown before the rebuild triggers? Even just a simple black screen with two sentences would suffice (after all, it is not a screen that should be encountered often.. and if this is not the case, then we'd have a bigger problem).

GIMP is not a neuroscience toolkit, which has different constraints. And the GIMP team is a team, not one single overworked person that has to deal with hundred-thousands of installations on three different operating systems, interacting with the system at often a way lower level than desktop software like GIMP. This is a pointless comparison that again sounds like backseat driving and entitlement to me.

Here's the thing. Do you seriously think I - after spending dozens to hundreds of hours on the text renderer implementation - never noticed that the cache rebuild times are annoying and an issue? Or that I wouldn't have already fixed it if this were as easy to fix as you seem to think, without creating various new and possibly worse problems? The reason it is the way it is, is because this is not a straightforward problem to fix. Nothing related to text rendering, especially on Windows, ever is. I decided, that the current state is the lesser of multiple evils for the time being. Especially given that users also can just follow our recommendation and use Linux, where one almost never encounters these problems, and fontconfig caches are automatically rebuild as needed at system installation/update/bootup time, so software like PTB doesn't have to deal with it. I don't have the time and resources to solve every possible problem on all operating systems, so I have to prioritize. And this is not a priority.

> Assuming there is even a way to know this ahead of time.

There isn't, to my knowledge, and I spent a lot of time searching through the docs and source code of libfontconfig. And there is also no way to know how long the rebuild will take. And just flushing some slide of text in the middle of a running visual stimulation paradigm on the first invocation of Screen('DrawText'), even for only a few milliseconds, would certainly not be appreciated by most experimenters, but badly disruptive and potentially confusing. Then we'd get bug reports about that. So you'd need multi-threading and timeout mechanisms that only show your text after say over 10 seconds have elapsed for a long cache rebuild.

And then a way to go back to the original visual stimulus once the rebuild is done.

Also the rendering and display code is in Screen, but the fontconfig handling is in an external plugin that only has limited facilities for graphics or display control, e.g., no way to trigger a display update. And your magic slide would also have to deal with non bog-standard display configurations, e.g., high precision framebuffers, HDR displays, stereoscopic and geometry corrected modes, special visual stimulators like from CRS and VPixx and others, Virtual Reality display HMD's, the alternative Vulkan display backend ... to display correctly, or at least not badly screw up those non-standard modes.

You can also not easily do this during Screen startup, when the normal welcome screen shows, as Screen supports in principle many different text renderer implementations, dynamically selectable by the user script, encapsulated in plugins that only get loaded at first use, and these plugins may have totally different init requirements. Some may need fontconfig, others don't. Only the plugin itself knows what it needs for bringup and breaking that separation between Screen and the plugins seems unwise and fraught with trouble as well.

The GIMP team probably "just" needed some async callback or background thread and all the progress bar and feedback code they already have for various other parts of GIMP, when some visual progress indicator needs to be displayed on slow operations. An interactive GUI software like GIMP is probably well equipped in its UI toolkit to do such things. It is a more simple problem.

For PTB and its use context, this is not an easy problem, not even if I were paid for solving it. But feel free to impress me with a very well thought out pull request that deals with all known and probably various so far unknown problems, without causing new problems, or disruption of any experiments or backwards compatibility or flexibility if it matter that much to you. Backseat driving is easy, doing is hard and time consuming.

**Comment by mirh** [comment]
> If it triggers every day, there might be an issue with the fontconfig cache.

Sure, see also my links in the OP. 
Yet this isn't advertised anywhere prominent in the program, despite what I would argue is a rather frequent occurrence. 
And few people have as much time to waste as myself, so that's why a shortcut to `help DrawTextPlugin` would be golden. 
(in fact, now that I think to it.. getting people to fix the underlying problem is probably better than handling its symptoms)

> Maybe if that entire lab is getting pissed

Uhm, lowkey pissed I meant? Not full-blown anger. 
(btw I meant lab as in "room with computers full of students", not lab as in "workplace with phds")

> First of all, I don't care about such little "cheap quirky voodoo" as you state it in this nice entitled sounding way,

I'm not entitled to anything, I don't know how you can have this reading. 
I just said what's the *feeling* that you get when "unexplainable random stuff" starts to happen in front of you.
(despite what is eventually the hard *fact* that the backend is continental-crust-solid, with insane actual precision)

> This is a pointless comparison that again sounds like backseat driving and entitlement to me.

It wasn't a comparison. It was just a stupid observation that 100 lines of code seemed trivial. 
My "laugh" was turned to myself and my naivety. 

> Here's the thing. Do you seriously think I - after spending dozens to hundreds of hours on the text renderer implementation - never noticed that the cache rebuild times are annoying and an issue?

Not never, but possibly rarely enough not to have ever come across the forefront of your mind?
And especially if perhaps windows isn't even your first rated/favourite platform (comprehensibly). 

> Or that I wouldn't have already fixed it if this were as easy to fix as you seem to think, without creating various new and possibly worse problems?

I see issues as reminders, not as demands. 
It's almost a year this (and others issues, like the fontconfig one) is open and I never even just nudged anybody. 

> There isn't, to my knowledge, and I spent a lot of time searching through the docs and source code of libfontconfig.

Good to know, so even my plan B (C?) sucks. 
Is it possible to at least detect whether a rebuild happened, that is, after the fact?
In that case, the event could at least be pointed out in a PTB-INFO line (for users to more easily know/search should they get annoyed).

**Comment by kleinerm** [comment]
> I'm not entitled to anything, I don't know how you can have this reading.
I just said what's the feeling that you get when "unexplainable random stuff" starts to happen in front of you.
(despite what is eventually the hard fact that the backend is continental-crust-solid, with insane actual precision)

"On the Internet, nobody knows you are a cat." It is not so much what you write, but how you write it, that comes over at least to me as backseat driving and some entitlement. Communication style matters, especially when dealing with open-source maintainers which are often massively overworked and close to burnout. I guess you are not a software developer or maintainer, so you don't have a good feeling for how much work some of your suggestions, or "100 lines of code that seem trivial" imply. Anyhow, moving on...

> Good to know, so even my plan B (C?) sucks.
Is it possible to at least detect whether a rebuild happened, that is, after the fact?
In that case, the event could at least be pointed out in a PTB-INFO line (for users to more easily know/search should they get annoyed).

You can't detect it reliably with reasonable effort. However, I added something like that for the upcoming release: It will time how long drawtext plugin initialization takes (of which a potential fontconfig cache rebuild could be one component), and if it takes more than 5 seconds, an info message will be printed, following your suggestion. I don't know if this will really help anybody, because first these cache issues seem to be only a MS-Windows thing, and then they probably don't happen that often anymore. But also, because many people don't read whats printed in the windows by PTB. It probably can't hurt though...

To give you an idea, this is a really trivial 10 lines code change (https://github.com/kleinerm/Psychtoolbox-3/commit/eefedba8ee232d1800075649c05a54463e76098e), which still took over 3 hours of work time to do and validate and tune, leading to multiple hundred euros of cost for the work time. A fontconfig cache rebuild on a cold booted MS-Windows with a 8 GB RAM machine and Matlab and Octave running, on a slow mechanical disc drive takes about 25 seconds on my machine, but less than a second on a machine with SSD. And with repeated rebuilds we are way below one second as well, if the OS has all the relevant font files still in its filesystem cache.

I think the main problem is a fontconfig cache rebuild on a cold booted / low on RAM system with old mechanical spinning disc drives (like my super slow Win-10 development machine), where this can take dozens of seconds. But then most people won't notice as this happens during PTB install/updates. And otherwise the real danger is impatient people killing Matlab/Octave from the task manager if it appears to hang for more than a few seconds, so no feedback about troubleshooting is possible. So this feedback would only help on slow machines, with fontconfig problems, and with users that are patient enough to wait a minute and have the gift of reading.
